tools/h5part_reader.py

```diff
@@ -0,0 +1,134 @@
+#!/bin/env python
+"""Script to read CPO data from fluidity h5part files.
+
+CPO data is saved to a NPZ file which can be read by `pydrex.mineral.Mineral.from_file`.
+The data is extracted from particles with a `scalar_attribute_array` named 'CPO',
+and saved to the NPZ file under fields named with the postfix set to the particle ID.
+For example, the fields 'meta_1', 'fractions_1' and 'orientations_1' store data for the
+particle with id 1. The particle positions are stored in the additional fields `x`, `y`
+and `z`.
+
+"""
+import argparse
+import io
+import os
+from zipfile import ZipFile
+
+import h5py
+import numpy as np
+
+from pydrex import minerals as _minerals
+
+
+def _get_args() -> argparse.Namespace:
+    description, epilog = __doc__.split(os.linesep + os.linesep, 1)
+    parser = argparse.ArgumentParser(description=description, epilog=epilog)
+    parser.add_argument(
+        "input",
+        help="input file (.h5part)",
+    )
+    parser.add_argument(
+        "-p",
+        "--phase",
+        help="mineral phase, default: olivine",
+        default="olivine",
+    )
+    parser.add_argument(
+        "-f",
+        "--fabric",
+        help="olivine fabric type, default: A, use 'N' for enstatite",
+        default="A",
+    )
+    parser.add_argument(
+        "-o",
+        "--output",
+        help="filename for the output NPZ file",
+    )
+    parser.add_argument(
+        "-n",
+        "--ngrains",
+        type=int,
+        help="number of grains in the polycrystal",
+        required=True,
+    )
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    args = _get_args()
+
+    if not args.input.endswith(".h5part"):
+        raise ValueError(
+            f"expected input file with .h5part extension, not {args.input}"
+        )
+    outfile = args.input[:-6] + "npz"  # Replace `.h5part` with `.npz`.
+    if type(args.output) == str and args.output != "":
+        if args.output.endswith(".npz"):
+            outfile = args.output
+        else:
+            raise ValueError(f"can only save to NPZ format, not {args.output}")
```

Comment:
Here, you first define `outfile`, but then you only proceed if two conditions are met, in which case you define `outfile` again.

Reply:
This gives the option to use `-o <outfile>` to explicitly name the NPZ output file, otherwise it just uses the input file name but with the extension changed to `.npz`. Maybe there's a cleaner way but that's why I'm redefining: default is <input_file>.npz but it can be overriden.